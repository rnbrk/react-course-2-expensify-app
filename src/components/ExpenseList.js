import React from 'react';
import { connect } from 'react-redux';
import ExpenseListItem from './ExpenseListItem';
import selectExpenses from '../selectors/expenses';

export const ExpenseList = props => (
  <div>
    {props.expenses.length === 0 ? (
      <p>No expenses</p>
    ) : (
      props.expenses.map(expense => <ExpenseListItem key={expense.id} {...expense} />)
    )}
    {}
  </div>
);

// MAP REDUX STATE TO THE PROPS (CALLBACK)
// WILL UPDATE EVERY TIME SOMETHING CHANGES TO THE STATE
const mapStateToProps = state => ({
  expenses: selectExpenses(state.expenses, state.filters)
});

// CONNECT TO THE STORE WITH
// FUNCTION (returns)-> HIGHER ORDER COMPONENT (returns)-> REACT COMPONENT
export default connect(mapStateToProps)(ExpenseList);
