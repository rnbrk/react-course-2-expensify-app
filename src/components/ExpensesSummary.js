import React from 'react';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import numeral from 'numeral';
import selectExpensesTotal from '../selectors/selectExpensesTotal';
import selectExpenses from '../selectors/expenses';

export const ExpensesSummary = ({ expenseCount, expensesTotal }) => {
  const expenseWord = expenseCount === 1 ? 'expense' : 'expenses';
  const formattedExpensesTotal = numeral(expensesTotal / 100).format('$0,0.00');

  return (
    <div className="page-header">
      <div className="content-container">
        <h1 className="page-header__title">
          Viewing <span>{expenseCount}</span> {expenseWord} totaling{' '}
          <span>{formattedExpensesTotal}</span>
          <div className="page-header__actions">
            <Link className="button" to="/create">
              Add expense
            </Link>
          </div>
        </h1>
      </div>
    </div>
  );
};

const mapStateToProps = state => {
  const visibleExpenses = selectExpenses(state.expenses, state.filters);
  return {
    expenseCount: visibleExpenses.length,
    expensesTotal: selectExpensesTotal(visibleExpenses)
  };
};

export default connect(mapStateToProps)(ExpensesSummary);
