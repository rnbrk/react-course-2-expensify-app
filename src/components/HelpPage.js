import React from 'react';

const HelpPage = () => (
  <div>The Help component!</div>
);

export default HelpPage;
