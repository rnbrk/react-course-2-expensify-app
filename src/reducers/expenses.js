const expensesReducer = (state = [], action) => {
  switch (action.type) {
    case 'ADD_EXPENSE':
      return [
        // Copies expenses array and adds an extra expense
        ...state,
        action.expense
      ];
    case 'REMOVE_EXPENSE':
      return state.filter(({ id }) => id !== action.id);
    case 'EDIT_EXPENSE':
      return state.map(expense =>
        expense.id === action.id
          ? {
              ...expense,
              ...action.updates
            }
          : expense
      );
    case 'SET_EXPENSES':
      return [...action.expenses];
    default:
      return state;
  }
};

export default expensesReducer;
