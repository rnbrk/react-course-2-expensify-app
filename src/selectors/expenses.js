import moment from 'moment';

// Filters and sorts the expenses according to the supplied filters
const getVisibleExpenses = (expenses, { text, sortBy, startDate, endDate }) =>
  expenses
    .filter(expense => {
      const createdAtMoment = moment(expense.createdAt);
      const startDateMatch = startDate ? startDate.isSameOrBefore(createdAtMoment, 'day') : true;
      const endDateMatch = endDate ? endDate.isSameOrAfter(createdAtMoment, 'day') : true;
      const textMatch = expense.description.toLowerCase().includes(text.toLowerCase());

      return startDateMatch && endDateMatch && textMatch;
    })
    .sort((a, b) => {
      if (sortBy === 'date') {
        return b.createdAt - a.createdAt;
      }
      if (sortBy === 'amount') {
        return b.amount - a.amount;
      }
    });

export default getVisibleExpenses;
