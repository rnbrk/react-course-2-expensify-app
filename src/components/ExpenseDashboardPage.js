import React from 'react';
import ExpenseListConnect from './ExpenseList';
import ExpenseListFilters from './ExpenseListFilters';
import ExpensesSummary from './ExpensesSummary';

const ExpenseDashboardPage = () => (
  <div>
    <ExpensesSummary />
    <ExpenseListFilters />
    <ExpenseListConnect />
  </div>
);

export default ExpenseDashboardPage;
