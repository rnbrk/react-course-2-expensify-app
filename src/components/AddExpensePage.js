import React from 'react';
import { connect } from 'react-redux';
import { addExpense } from '../actions/expenses';
import ExpenseForm from './ExpenseForm';

export class AddExpensePage extends React.Component {
  onSubmit = expense => {
    // dispatch(addExpense(expense));
    this.props.addExpense(expense); // Replaces line above
    this.props.history.push('/');
  };

  render() {
    return (
      <div>
        <div>Add expense</div>
        <ExpenseForm onSubmit={this.onSubmit} />
      </div>
    );
  }
}

// This makes sure we can test this component easier
const mapDispatchToProps = dispatch => ({
  addExpense: expense => dispatch(addExpense(expense))
});

export default connect(
  undefined,
  mapDispatchToProps
)(AddExpensePage);
