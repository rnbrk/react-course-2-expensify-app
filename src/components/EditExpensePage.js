import React from 'react';
import { connect } from 'react-redux';
import { startEditExpense, startRemoveExpense } from '../actions/expenses';
import ExpenseForm from './ExpenseForm';

export class EditExpensePage extends React.Component {
  onSubmit = expense => {
    this.props.startEditExpense(this.props.expense.id, expense);
    this.props.history.push('/');
  };

  onClick = () => {
    this.props.startRemoveExpense(this.props.expense.id);
    this.props.history.push('/');
  };

  render() {
    return (
      <div>
        <ExpenseForm expense={this.props.expense} onSubmit={this.onSubmit} />
        <button type="button" onClick={this.onClick}>
          Remove
        </button>
      </div>
    );
  }
}

const mapDispatchToProps = dispatch => ({
  startEditExpense: (id, expense) => dispatch(startEditExpense(id, expense)),
  startRemoveExpense: id => dispatch(startRemoveExpense(id))
});

const mapStateToProps = (state, props) => ({
  expense: state.expenses.find(expense => expense.id === props.match.params.id)
});

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(EditExpensePage);
