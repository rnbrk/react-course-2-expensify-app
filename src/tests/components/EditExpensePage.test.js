import React from 'react';
import { shallow } from 'enzyme';
import { EditExpensePage } from '../../components/EditExpensePage';
import expenses from '../fixtures/expenses';

let editExpenseSpy, expense, history, removeExpenseSpy, wrapper;

beforeEach(() => {
  editExpenseSpy = jest.fn();
  expense = expenses[2];
  history = { push: jest.fn() };
  removeExpenseSpy = jest.fn();
  wrapper = shallow(
    <EditExpensePage
      editExpense={editExpenseSpy}
      history={history}
      removeExpense={removeExpenseSpy}
      expense={expense}
    />
  );
});

test('Should render EditExpensePage', () => {
  expect(wrapper).toMatchSnapshot();
});

test('Should handle editExpense', () => {
  wrapper.find('ExpenseForm').prop('onSubmit')(expense);
  expect(editExpenseSpy).toHaveBeenLastCalledWith(expense.id, expense);
  expect(history.push).toHaveBeenLastCalledWith('/');
});

test('Should handle removeExpense', () => {
  wrapper.find('button').simulate('click');
  expect(removeExpenseSpy).toHaveBeenLastCalledWith(expense.id);
  expect(history.push).toHaveBeenLastCalledWith('/');
});
