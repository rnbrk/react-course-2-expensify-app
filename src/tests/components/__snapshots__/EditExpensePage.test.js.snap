// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render EditExpensePage 1`] = `
<div>
  <ExpenseForm
    expense={
      Object {
        "amount": 5999,
        "createdAt": 345600000,
        "description": "Videogame",
        "id": "3",
        "note": "",
      }
    }
    onSubmit={[Function]}
  />
  <button
    onClick={[Function]}
    type="button"
  >
    Remove
  </button>
</div>
`;
